server/sonar-server/src/main/java/org/sonar/server/permission/GroupPermissionChanger.java

```diff
@@ -48,6 +55,53 @@ public boolean apply(DbSession dbSession, GroupPermissionChange change) {
     }
   }
 
+  private boolean isImplicitlyAlreadyDone(GroupPermissionChange change) {
+    return change.getProjectId()
+      .map(projectId -> isImplicitlyAlreadyDone(projectId, change))
+      .orElse(false);
+  }
+
+  private boolean isImplicitlyAlreadyDone(ProjectId projectId, GroupPermissionChange change) {
+    return isAttemptToAddPublicPermissionToPublicComponent(change, projectId)
+      || isAttemptToRemovePermissionFromAnyoneOnPrivateComponent(change, projectId);
+  }
+
+  private static boolean isAttemptToAddPublicPermissionToPublicComponent(GroupPermissionChange change, ProjectId projectId) {
+    return !projectId.isPrivate()
+      && change.getOperation() == ADD
+      && ProjectPermissions.PUBLIC_PERMISSIONS.contains(change.getPermission());
+  }
+
+  private static boolean isAttemptToRemovePermissionFromAnyoneOnPrivateComponent(GroupPermissionChange change, ProjectId projectId) {
+    return projectId.isPrivate()
+      && change.getOperation() == REMOVE
+      && ProjectPermissions.PUBLIC_PERMISSIONS.contains(change.getPermission());
```

Comment:
According to the current implementation, this method should be called `isAttemptToRemovePublicPermissionToPrivateComponent`.

:exclamation: Change `&& ProjectPermissions.PUBLIC_PERMISSIONS.contains(change.getPermission());` to `&& change.getGroupIdOrAnyone().isAnyone();`